### ISSUE TITLE
QR decomposition and mutating arrays

### ISSUE THREAD

**Issue by hyun-wooh** [JuliaLang/julia#50609]
Hi. I'm trying to use pivoted QR decomposition in the loss function of ML and calculate its gradient.  

My code is

```Julia
using LinearAlgebra
using Zygote

# Define a simple function that uses qr
function my_function(A) 
    Q, R, P = qr(A, Val(true))
    return R^2
end

# Define an example input matrix
A = [1.0 2.0; 3.0 4.0]

# Compute the gradient of the sum of elements of my_function's output with respect to A using Zygote
gradient_value = Zygote.gradient(A -> sum(my_function(A)), A)
```

And the error message is 

```
Mutating arrays is not supported -- called setindex!(Matrix{Float64}, ...)
This error occurs when you ask Zygote to differentiate operations that change
the elements of arrays in place (e.g. setting values with x .= ...)
```

Even this happens for normal qr, but I couldn't see this with other decompositions, such as LU or cholesky, etc. Is there any reason that only QR has a different behavior?

**Comment by aravindh-krishnamoorthy** [comment]
This is because `ChainRules.jl` does not yet have rules for the QR decomposition. But, interested parties can implement it themselves very easily [as described here](https://rkube.github.io/jekyll/update/2021/07/06/backpropagating-through-qr.html) or raise an issue on https://github.com/JuliaDiff/ChainRules.jl. Perhaps it is better to close this issue on JuliaLang?
